astronomy/orbital_elements_body.hpp

```diff
@@ -0,0 +1,355 @@
+
+inline OrbitalElements::Interval<Angle>
+OrbitalElements::mean_argument_of_periapsis_range() const {
+  return mean_argument_of_periapsis_range_;
+}
+
+template<typename PrimaryCentred>
+std::vector<OrbitalElements::EquinoctialElements>
+OrbitalElements::OsculatingEquinoctialElements(
+    DiscreteTrajectory<PrimaryCentred> const& trajectory,
+    MassiveBody const& primary,
+    Body const& secondary) {
+  DegreesOfFreedom<PrimaryCentred> const primary_dof{
+      PrimaryCentred::origin, Velocity<PrimaryCentred>{}};
+  std::vector<EquinoctialElements> result;
+  for (auto it = trajectory.Begin(); it != trajectory.End(); ++it) {
+    auto const osculating_elements =
+        KeplerOrbit<PrimaryCentred>(primary,
+                                    secondary,
+                                    it.degrees_of_freedom() - primary_dof,
+                                    it.time())
+            .elements_at_epoch();
+    double const& e = *osculating_elements.eccentricity;
+    Angle const& ϖ = *osculating_elements.longitude_of_periapsis;
+    Angle const& Ω = osculating_elements.longitude_of_ascending_node;
+    Angle const& M = *osculating_elements.mean_anomaly;
+    Angle const& i = osculating_elements.inclination;
+    double const tg_½i = Tan(i / 2);
+    double const cotg_½i = 1 / tg_½i;
+    result.push_back(
+        {/*.t = */ it.time(),
+         /*.a = */ *osculating_elements.semimajor_axis,
+         /*.h = */ e * Sin(ϖ),
+         /*.k = */ e * Cos(ϖ),
+         /*.λ = */ result.empty() ? ϖ + M : UnwindFrom(result.back().λ, ϖ + M),
+         /*.p = */ tg_½i * Sin(Ω),
+         /*.q = */ tg_½i * Cos(Ω),
+         /*.pʹ = */ cotg_½i * Sin(Ω),
+         /*.qʹ = */ cotg_½i * Cos(Ω)});
+  }
+  return result;
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements> const&
+OrbitalElements::osculating_equinoctial_elements() const {
+  return osculating_equinoctial_elements_;
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements> const&
+OrbitalElements::mean_equinoctial_elements() const {
+  return mean_equinoctial_elements_;
+}
+
+inline Time OrbitalElements::SiderealPeriod(
+    std::vector<EquinoctialElements> const& equinoctial_elements) {
+  Time const Δt =
+      equinoctial_elements.back().t - equinoctial_elements.front().t;
+  Instant const t0 = equinoctial_elements.front().t + Δt / 2;
+  Product<Angle, Square<Time>> ſ_λt_dt;
+
+  auto const first = equinoctial_elements.begin();
+  Time const first_t = first->t - t0;
+  Product<Angle, Time> previous_λt = first->λ * first_t;
+  for (auto previous = first, it = first + 1; it != equinoctial_elements.end();
+       previous = it, ++it) {
+    Time const t = it->t - t0;
+    auto const λt = it->λ * t;
+    Time const dt = it->t - previous->t;
+    ſ_λt_dt += (λt + previous_λt) / 2 * dt;
+    previous_λt = λt;
+  }
+  return 2 * π * Radian * Pow<3>(Δt) / (12 * ſ_λt_dt);
+}
+
+inline std::vector<OrbitalElements::EquinoctialElements>
+OrbitalElements::MeanEquinoctialElements(
```

Comment:
I have no idea what's happening in this function.  Comments would be good.

Reply:
Added comments and adapted the notation to match the comments.